File: DCP-1837/solution.swift
// Problem 1837 (Stable Marriage Problem)

import Foundation

struct StableMarriage {
    let menPreferences: [String: [String]]
    let womenPreferences: [String: [String]]
    
    func match() -> [String: String] {
        var freeMen = Set(menPreferences.keys)
        var currentMatches = [String: String]()
        var womenRankings: [String: [String: Int]] = [:]
        
        // Precompute rankings for O(1) lookup
        for (woman, prefs) in womenPreferences {
            womenRankings[woman] = [:]
            for (rank, man) in prefs.enumerated() {
                womenRankings[woman]?[man] = rank
            }
        }
        
        while let man = freeMen.first {
            let preferences = menPreferences[man] ?? []
            for woman in preferences {
                if !currentMatches.values.contains(woman) {
                    currentMatches[man] = woman
                    freeMen.remove(man)
                    break
                } else {
                    let currentMan = currentMatches.first { $0.value == woman }!.key
                    let currentRank = womenRankings[woman]?[currentMan] ?? 0
                    let newRank = womenRankings[woman]?[man] ?? 0
                    
                    if newRank < currentRank {
                        currentMatches[man] = woman
                        freeMen.remove(man)
                        freeMen.insert(currentMan)
                        break
                    }
                }
            }
        }
        
        return currentMatches
    }
}

File: DCP-1837/solution.kt
// Problem 1837

class StableMarriage(
    private val menPreferences: Map<String, List<String>>,
    private val womenPreferences: Map<String, List<String>>
) {
    fun match(): Map<String, String> {
        val freeMen = menPreferences.keys.toMutableSet()
        val currentMatches = mutableMapOf<String, String>()
        val womenRankings = womenPreferences.mapValues { (_, prefs) ->
            prefs.withIndex().associate { it.value to it.index }
        }
        
        while (freeMen.isNotEmpty()) {
            val man = freeMen.first()
            val preferences = menPreferences[man] ?: continue
            
            for (woman in preferences) {
                if (woman !in currentMatches.values) {
                    currentMatches[man] = woman
                    freeMen.remove(man)
                    break
                } else {
                    val currentMan = currentMatches.entries.first { it.value == woman }.key
                    val currentRank = womenRankings[woman]?.get(currentMan) ?: 0
                    val newRank = womenRankings[woman]?.get(man) ?: 0
                    
                    if (newRank < currentRank) {
                        currentMatches[man] = woman
                        freeMen.remove(man)
                        freeMen.add(currentMan)
                        break
                    }
                }
            }
        }
        
        return currentMatches
    }
}